### PR TITLE
chore(test): validate generated typings

### LINKED ISSUE #7695: Better coverage for gulp test.typings
Followup from https://github.com/angular/angular/issues/7656#issuecomment-198413877

Beta.10 was broken because our typings test didn't cover a simple scenario. We should have test coverage so we don't break releases ability to type-check.

One nice answer is to take all the examples that are in the repo, and use them as additional test cases in the test.typings setup in gulpfile.js

### PR DESCRIPTION
**Pull request by JaKXz**
Lots of issues were raised with recent releases because the generated typings were not checked for compilation errors (e.g. #7656).

Closes #7695.
- **Other information**:
  Thanks @nexus-uw for pairing with me! :tada: 

cc @alexeagle

### CODE REVIEW

**Comment by JaKXz** [comment]
@googlebot I signed it!

File: gulpfile.js [hunk 1]
```diff
@@ -984,18 +984,30 @@ gulp.task('static-checks', ['!build.tools'], function(done) {
 // Make sure the two typings tests are isolated, by running this one in a tempdir
 var tmpdir = path.join(os.tmpdir(), 'test.typings', new Date().getTime().toString());
 gulp.task('!pre.test.typings.layoutNodeModule', ['build.js.cjs'], function() {
-  return gulp.src(['dist/js/cjs/angular2/**/*', 'node_modules/rxjs/**'], {base: 'dist/js/cjs'})
+  return gulp.src(['dist/js/cjs/angular2/**/*', 'node_modules/rxjs/**/*'], {base: 'dist/js/cjs'})
       .pipe(gulp.dest(path.join(tmpdir, 'node_modules')));
 });
+
+gulp.task('!pre.test.typings.copyTools', function() {
+  return gulp.src(['modules/angular2/typings/!(es6-*)/*'], {base: 'modules/angular2/typings'})
```
Comment (alexeagle) [inline gulpfile.js]: we should explicitly list anything we copy in, rather than use exclusions.

File: gulpfile.js [hunk 2]
```diff
@@ -984,18 +984,30 @@ gulp.task('static-checks', ['!build.tools'], function(done) {
...
+      .pipe(gulp.dest(tmpdir));
+});
+
 gulp.task('!pre.test.typings.copyTypingsSpec', function() {
-  return gulp.src(['typing_spec/*.ts'], {base: 'typing_spec'}).pipe(gulp.dest(tmpdir));
```
Comment (alexeagle) [inline gulpfile.js]: are the files under typing_spec no longer used as a result of this change? delete them as well?

File: gulpfile.js [hunk 3]
```diff
@@ -984,18 +984,30 @@ gulp.task('static-checks', ['!build.tools'], function(done) {
 // Make sure the two typings tests are isolated, by running this one in a tempdir
 var tmpdir = path.join(os.tmpdir(), 'test.typings', new Date().getTime().toString());
 gulp.task('!pre.test.typings.layoutNodeModule', ['build.js.cjs'], function() {
-  return gulp.src(['dist/js/cjs/angular2/**/*', 'node_modules/rxjs/**'], {base: 'dist/js/cjs'})
+  return gulp.src(['dist/js/cjs/angular2/**/*', 'node_modules/rxjs/**/*'], {base: 'dist/js/cjs'})
       .pipe(gulp.dest(path.join(tmpdir, 'node_modules')));
 });
+
+gulp.task('!pre.test.typings.copyTools', function() {
```
Comment (alexeagle) [inline gulpfile.js]: these are not tools, but dependencies right?

File: modules/angular2/examples/core/pipes/ts/async_pipe/async_pipe_example.ts [hunk 4]
```diff
@@ -38,7 +38,7 @@ export class AsyncPipeExample {
 @Component({selector: "task-cmp", template: "Time: {{ time | async }}"})
 class Task {
   time = new Observable<number>((observer: Subscriber<number>) => {
-    setInterval(_ => observer.next(new Date().getTime()), 500);
+    setInterval(() => observer.next(new Date().getTime()), 500);
```
Comment (alexeagle) [inline modules/angular2/examples/core/pipes/ts/async_pipe/async_pipe_example.ts]: why is this change needed?

Comment (JaKXz) [inline modules/angular2/examples/core/pipes/ts/async_pipe/async_pipe_example.ts]: Because `_` was throwing a `tsc` error as an implicit `any`. Since it's not being used I think this reads better anyway.

File: gulpfile.js (hunk 3 above)
Comment (JaKXz) [inline gulpfile.js]: Good point, I'll change that.

File: gulpfile.js (hunk 2 above)
Comment (JaKXz) [inline gulpfile.js]: Sure!

File: modules/angular2/examples/facade/ts/async/observable.ts [hunk 5]
```diff
@@ -2,7 +2,7 @@
 import {Observable, Subscriber} from 'rxjs/Rx';
 var obs = new Observable<number>((obs: Subscriber<number>) => {
   var i = 0;
-  setInterval(_ => { obs.next(++i); }, 1000);
+  setInterval(() => { obs.next(++i); }, 1000);
 });
-obs.subscribe(i => console.log(`${i} seconds elapsed`));
```
Comment (alexeagle) [inline modules/angular2/examples/facade/ts/async/observable.ts]: was this an implicitAny error as well?

Comment (JaKXz) [inline modules/angular2/examples/facade/ts/async/observable.ts]: No - I'm a bit surprised it wasn't actually -

Comment (alexeagle) [inline modules/angular2/examples/facade/ts/async/observable.ts]: I'm surprised that it makes any difference whether a single parameter is wrapped in parens or not.
I think the style preference is to leave the LHS of the arrow bare (no parens) when it has a single arg like in this case.
(unless you find it actually is required for some reason)

File: gulpfile.js [hunk 6]
```diff
@@ -984,18 +984,36 @@ gulp.task('static-checks', ['!build.tools'], function(done) {
 // Make sure the two typings tests are isolated, by running this one in a tempdir
 var tmpdir = path.join(os.tmpdir(), 'test.typings', new Date().getTime().toString());
 gulp.task('!pre.test.typings.layoutNodeModule', ['build.js.cjs'], function() {
-  return gulp.src(['dist/js/cjs/angular2/**/*', 'node_modules/rxjs/**'], {base: 'dist/js/cjs'})
+  return gulp.src(['dist/js/cjs/angular2/**/*', 'node_modules/rxjs/**/*'], {base: 'dist/js/cjs'})
```
Comment (alexeagle) [inline gulpfile.js]: just curious, did the extra star actually make a difference?

File: gulpfile.js [hunk 7]
```diff
@@ -984,18 +984,36 @@ gulp.task('static-checks', ['!build.tools'], function(done) {
...
       .pipe(gulp.dest(path.join(tmpdir, 'node_modules')));
 });
+
+gulp.task('!pre.test.typings.copyDeps', function() {
+  return gulp.src(
+                 [
+                   'modules/angular2/typings/angular-protractor/*.ts',
+                   'modules/angular2/typings/jasmine/*.ts',
+                   'modules/angular2/typings/node/*.ts',
```
Comment (alexeagle) [inline gulpfile.js]: hmm, this is smelly, it allows us to publish things with arbitrary node dependency everywhere, even though users should not need node typings except in very specific usages (eg. parse5 tests without a browser)
Is this really needed to make the examples compile?

File: typing_spec/router_spec.ts [hunk 8]
```diff
@@ -1,28 +0,0 @@
-import {Component} from 'angular2/core';
-import {bootstrap} from 'angular2/platform/browser';
-import {RouteConfig, ROUTER_DIRECTIVES, ROUTER_PROVIDERS} from 'angular2/router';
```
Comment (alexeagle) [inline typing_spec/router_spec.ts]: we do have some examples that still provide router API coverage in the test.typings, right?

File: gulpfile.js (hunk 6 above)
Comment (JaKXz) [inline gulpfile.js]: It was added to be consistent with the other globs and pull everything reliably. `/**` would throw errors that were solved by adding the extra star.

File: gulpfile.js (hunk 7 above)
Comment (JaKXz) [inline gulpfile.js]: Good catch, got rid of node - everything else is definitely needed.

File: typing_spec/router_spec.ts (hunk 8 above)
Comment (JaKXz) [inline typing_spec/router_spec.ts]: Yep, I believe so!

File: modules/angular2/examples/facade/ts/async/observable.ts (hunk 5 above)
Comment (JaKXz) [inline modules/angular2/examples/facade/ts/async/observable.ts]: Sure, I'll revert it back - is a whole new commit preferred?

Comment (JaKXz) [inline modules/angular2/examples/facade/ts/async/observable.ts]: *edit:* or just for explicitness I can specify the type on `i`. I'm fine with either, what would you suggest? :)

Comment (alexeagle) [inline modules/angular2/examples/facade/ts/async/observable.ts]: Let's keep the change specific to the task at hand, so I'd prefer reverting unneeded bits.

**Comment by alexeagle** [comment]
Can you re-base and squash all the commits together?

**Comment by alexeagle** [comment]
Or two commits could be okay too if you want to do the git operations needed (one for fixing the examples for noImplicitAny, one for changing the test)

File: typing_spec/router_spec.ts (hunk 8 above)
Comment (nexus-uw) [inline typing_spec/router_spec.ts]: https://github.com/angular/angular/tree/master/modules/angular2/examples/router/ts covers a lot of it.

**Comment by JaKXz** [comment]
Thanks for the feedback @alexeagle, and the nice thing about this is that as examples are added, coverage increases with each one.

**Comment by alexeagle** [comment]
Yes! Much better

**Comment by JaKXz** [comment]
Sorry @alexeagle @mhevery I should've put "closes #7695" in the commit message (unless you want to double check that later on), I wasn't aware the PR would be merged differemtly.
